data_model/primitives/src/conststr.rs

```diff
@@ -0,0 +1,545 @@
+//! Const-string related implementation and structs.
+#[cfg(not(feature = "std"))]
+use alloc::{boxed::Box, str::from_boxed_utf8_unchecked, string::String, vec::Vec};
+use core::{
+    cmp::{Eq, Ord, Ordering, PartialEq, PartialOrd},
+    convert::{AsRef, From, TryFrom},
+    fmt::{Debug, Display, Formatter, Result as FmtResult},
+    hash::{Hash, Hasher},
+    mem::{size_of, ManuallyDrop},
+    ops::Deref,
+};
+#[cfg(feature = "std")]
+use std::str::from_boxed_utf8_unchecked;
+
+use iroha_schema::{IntoSchema, MetaMap};
+use parity_scale_codec::{Encode, Output};
+use serde::{Serialize, Serializer};
+
+const MAX_INLINED_STRING_LEN: usize = 2 * size_of::<usize>() - 1;
+
+/// Immutable inlinable string.
+/// Strings shorter than 15/7/3 bytes (depending on architecture 64/32/16 bit pointer width) are inlined.
+#[derive(Clone)]
+pub struct ConstString(ConstStringData);
+
+impl ConstString {
+    /// Returns the length of this [`ConstString`], in bytes.
+    #[inline]
+    pub fn len(&self) -> usize {
+        self.0.len()
+    }
+
+    /// Returns `true` if [`ConstString`] is empty.
+    #[inline]
+    pub fn is_empty(&self) -> bool {
+        self.0.len() == 0
+    }
+
+    /// Returns `true` if [`ConstString`] data is inlined.
+    #[inline]
+    pub const fn is_inlined(&self) -> bool {
+        self.0.is_inlined()
+    }
+
+    /// Construct empty [`ConstString`].
+    #[inline]
+    pub const fn new() -> Self {
+        Self(ConstStringData::new())
+    }
+}
+
+impl Deref for ConstString {
```

Comment:
check out `derive_more::Deref`

Reply:
Not a big deal

Reply:
So when you `derive` instead of implementing, you're communicating to the reader of the code that there's nothing interesting going on here. That yes, this object does implement this interface, but that there's nothing interesting in how it implements. 

In this particular case the `Deref` is straightforwardly the default. If you `derive(Deref)` you will communicate that there's nothing in there, even though the size of the code may not reduce, you will have a much clearer picture just looking at the struct definition.

Reply:
Deriving `Deref` requires inner union `ConstStringData` to be public, so i think we should leave it as is.

Reply:
has the situation changed now?

Reply:
not realy, beacuse we still can't derive `Deref` implementation and need to do it manually